zebra-consensus/src/primitives/groth16.rs

```diff
@@ -9,120 +9,117 @@ use std::{
-        let Item {
-            proof,
-            public_inputs,
-        } = self;
-
-        bellman::groth16::verify_proof(pvk, &proof, &public_inputs)
-    }
-}
-
-impl<E: MultiMillerLoop> From<(&Proof<E>, &[E::Fr])> for Item<E> {
-    fn from((proof, public_inputs): (&Proof<E>, &[E::Fr])) -> Self {
-        (proof.clone(), public_inputs.to_owned()).into()
-    }
-}
-
-impl<E: MultiMillerLoop> From<(Proof<E>, Vec<E::Fr>)> for Item<E> {
-    fn from((proof, public_inputs): (Proof<E>, Vec<E::Fr>)) -> Self {
-        Self {
-            proof,
-            public_inputs,
-        }
-    }
-}
-
-#[derive(Default)]
-struct Batch {
-    queue: Vec<Item<Bls12>>,
-}
-
-impl Batch {
-    fn queue(&mut self, item: Item<Bls12>) {
-        self.queue.push(item);
-    }
-
-    fn verify<R: RngCore + CryptoRng>(
-        self,
-        _rng: R,
-        pvk: &PreparedVerifyingKey<Bls12>,
-    ) -> Result<(), VerificationError> {
-        for item in self.queue {
-            item.verify_single(pvk)?;
-        }
-
-        Ok(())
-    }
-}
-
-// === TEMPORARY BATCH BELLMAN SUBSTITUTE END ===
-
-// /// A Groth16 verification item, used as the request type of the service.
-// pub type Item = batch::Item<Bls12>;
-
-/// Groth16 signature verifier service
-#[derive(Clone, Debug)]
-pub struct Verifier {
-    inner: Fallback<tower_batch::Batch<VerifierImpl, Item<Bls12>>, FallbackVerifierImpl>,
-}
-
-impl Verifier {
-    /// Constructs a new verifier.
-    pub fn new(pvk: &'static PreparedVerifyingKey<Bls12>) -> Self {
-        let verifier_impl = VerifierImpl::new(pvk);
-        let fallback_impl = FallbackVerifierImpl::new(pvk);
-
-        let max_items = super::MAX_BATCH_SIZE;
-        let max_latency = super::MAX_BATCH_LATENCY;
-
-        let inner = tower_batch::Batch::new(verifier_impl, max_items, max_latency);
-        let inner = Fallback::new(inner, fallback_impl);
-
-        Self { inner }
-    }
-}
-
-impl Service<Item<Bls12>> for Verifier {
-    type Response = ();
-    type Error = BoxError;
-    type Future = Pin<Box<dyn Future<Output = Result<(), BoxError>> + Send + 'static>>;
-
-    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
-        self.inner.poll_ready(cx)
-    }
-
-    fn call(&mut self, req: Item<Bls12>) -> Self::Future {
-        use futures::FutureExt;
-        self.inner.call(req).boxed()
+/// Global batch verification context for Groth16 proofs of Spend statements.
+///
+/// This service transparently batches contemporaneous proof verifications,
+/// handling batch failures by falling back to individual verification.
+///
+/// Note that making a `Service` call requires mutable access to the service, so
+/// you should call `.clone()` on the global handle to create a local, mutable
+/// handle.
+pub static SPEND_VERIFIER: Lazy<
+    Fallback<Batch<Verifier, Item>, ServiceFn<fn(Item) -> Ready<Result<(), VerificationError>>>>,
+> = Lazy::new(|| {
```

Comment:
we should not make this handle public if we want to enforce cloning, and instead expose a pub method that calls clone and returns that handle.

Reply:
We should make a similar change to our other batch verifiers, but that can be done in a follow-up ticket.

Reply:
Yeah we now have 4, going to be 5 batch verifier services that are currently clone()'d inside `transaction::Verifier::call()`, `script_verifier` is assigned to the Verifier instance on construction and later clone'd, the others are referenced and clone'd only within `call`. I think all are currently `static` besides `script::Verifier`. Are we doing this right?

Reply:
I think consistency is important, but we were always going to have a lot of batch verifiers, so having 5 doesn't worry me.

Reply:
To be specific: let's access, store, and clone them all the same way. Clones should be cheap.